src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -530,27 +531,35 @@ private static TypeSymbol MakeMergedTupleType(ArrayBuilder<DeconstructionVariabl
             return new BoundDeconstructionAssignmentStep(node, op, outputPlaceholder);
         }
 
-        private static ImmutableArray<BoundExpression> FlattenDeconstructVariables(ArrayBuilder<DeconstructionVariable> variables)
-        {
-            var builder = ArrayBuilder<BoundExpression>.GetInstance(variables.Count);
-            FlattenDeconstructVariables(variables, builder);
-
-            return builder.ToImmutableAndFree();
-        }
-
-        private static void FlattenDeconstructVariables(ArrayBuilder<DeconstructionVariable> variables, ArrayBuilder<BoundExpression> builder)
+        private BoundTupleLiteral DeconstructVariablesAsTuple(SyntaxNode syntax, ArrayBuilder<DeconstructionVariable> variables)
         {
+            int count = variables.Count;
+            var argBuilder = ArrayBuilder<BoundExpression>.GetInstance(count);
+            var elementsBuilder = ArrayBuilder<TypeSymbol>.GetInstance(count);
+            var locationsBuilder = ArrayBuilder<Location>.GetInstance(count);
             foreach (var variable in variables)
             {
                 if (variable.HasNestedVariables)
                 {
-                    FlattenDeconstructVariables(variable.NestedVariables, builder);
+                    var nestedTuple = DeconstructVariablesAsTuple(variable.Syntax, variable.NestedVariables);
+                    argBuilder.Add(nestedTuple);
+                    elementsBuilder.Add(nestedTuple.Type);
                 }
                 else
                 {
-                    builder.Add(variable.Single);
+                    var single = variable.Single;
+                    argBuilder.Add(single);
+                    elementsBuilder.Add(single.Type);
                 }
+                locationsBuilder.Add(variable.Syntax.Location);
             }
+
+            var type = TupleTypeSymbol.Create(syntax.Location,
+                elementsBuilder.ToImmutableAndFree(), locationsBuilder.ToImmutableAndFree(),
+                elementNames: default(ImmutableArray<string>), compilation: this.Compilation, shouldCheckConstraints: false);
```

Comment:
I'll add a comment.
This makes me think we could construct this final tuple only once. I'll file an issue.